llvm/lib/Target/AIE/AIEBaseAliasAnalysis.cpp

```diff
@@ -117,13 +152,286 @@ static const Value *getUnderlyingObjectAIE(const Value *V) {
+          }
+        }
+      } else {
+        return false; // CountA is not PHI.
+      }
+    }
+    return true;
+  }
+
+  bool notOverlap(const IntrinsicChainInfo &Other) const {
+    // Different number of updates in the chain.
+    // * but maybe with consistent updates (same steps)
+    if (!hasSameTotalNumberOfCalls(Other))
+      return false;
+
+    // Different parameters.
+    if (!hasSameParameters(Other))
+      return false;
+
+    // Every counter must start equally.
+    if (!hasSameStartingCounters(Other))
+      return false;
+
+    // Check for different steps of calculation.
+    if (hasSameNumberOfCalls(Other))
+      return false;
+
+    return true;
+  }
+};
+
+static IntrinsicArgsInfo collectIntrinsicOperands(const Value *V) {
+  const auto *Extract = dyn_cast<ExtractValueInst>(V);
+  assert(Extract && "Not an Extract");
+  const auto *IntrinsicCall =
+      dyn_cast<CallBase>(Extract->getAggregateOperand());
+  assert(IntrinsicCall && "Not an Intrinsic");
+  Intrinsic::ID ID = IntrinsicCall->getIntrinsicID();
+  IntrinsicArgsInfo Args;
+  unsigned InPtrIdx;
+  isAIEPtrAddIntrinsic(ID, InPtrIdx);
+  // Necessary operands to disambiguate.
+  // Constants.
+  for (unsigned Op : getAddIntrinsicOps(ID)) {
+    Args.ArgSet.push_back(IntrinsicCall->getArgOperand(Op));
+  }
+  // Counters.
+  for (unsigned Op : getAddIntrinsicCounterOps(ID)) {
+    Args.CounterArgSet.push_back(IntrinsicCall->getArgOperand(Op));
+  }
+  return Args;
+}
+
+// We can reach a pointer definition starting from a counter.
+static bool isDirectReachable(SmallVector<const Value *, 5> CounterArgSet,
+                              const Value *V) {
+
+  for (auto Counter : CounterArgSet) {
+    if (V != lookThroughAIEAddressingModes(Counter))
+      return false;
+  }
+
+  return true;
+}
+
+/// This function returns some tracking information related to the upward path
+/// that starts from a Value *TargetPointer to the Phi node that firstly defines
+/// a pointer value to the same ambiguous memory. We start the search from the
+/// last update of the pointer.
+std::optional<IntrinsicChainInfo>
+trackIntrinsicChain(const Value *LastUpdate, const Value *TargetPointer) {
+  IntrinsicChainInfo ChainInfo;
+  // We count backwards.
+  unsigned LastCalls = 0;
+  // Let's expose the real pointer.
+  TargetPointer = skipCast(TargetPointer);
+  if (!LastUpdate->getType()->isPointerTy())
+    return std::nullopt;
+  for (unsigned Count = 0; Count < BaseObjectSearchLimit; ++Count) {
+    if (auto *GEP = dyn_cast<GEPOperator>(LastUpdate)) {
+      if (GEP->hasAllZeroIndices()) {
+        // The result pointer and the first operand have the same value
+        LastUpdate = GEP->getPointerOperand();
+      } else {
+        return std::nullopt; // No pointer arithmetic is allowed.
+      }
+    } else if (Operator::getOpcode(LastUpdate) == Instruction::BitCast ||
+               Operator::getOpcode(LastUpdate) == Instruction::AddrSpaceCast) {
```

Comment:
SuperNit: That seems a bit redundant with `skipCast`. Maybe we could call `LastUpdate = skipCast(LastUpdate)` at every iteration? Then I guess we only need `if (!LastUpdate->getType()->isPointerTy()) { ... }`